README.md
# wuy 

If you have made a _super python tool_, and you want to add a GUI but without using qt/gtk/tk/wx/kivy/toga : you are in the right place ! Just re-use the local browser (chrome in app mode) as GUI, drived by **wuy**.

<p align="center">
    <img src="https://github.com/manatlan/wuy/blob/master/wiki/capture.png"/><br/>
    <a href="https://github.com/manatlan/wuy/blob/master/web/askName.html">html</a> / 
    <a href="https://github.com/manatlan/wuy/blob/master/an_app2.py">python</a>
</p>

Basically, it's a python module which will act as a web server (http & websocket) and will open/manage a window for you ; providing a simple framework (js/py3 helpers).

It lets you create your GUI with yours web skills (html/js/css or any js frameworks/libs), and re-use the power of python on server side.

Technically, it's a python3 module, using asyncio and the marvellous [aiohttp](https://aiohttp.readthedocs.io/en/stable/), and (if present) the [uvloop](https://magic.io/blog/uvloop-blazing-fast-python-networking/), for the full speed ! The http server is here to serve static content (html, js, images ...). The websocket is here to simplify the communication (sync/async) between the window & the server. (rpc method to communicate from client to server, pubsub mechanism to communicate from server to clients, or from client to clients)

In **app/window mode** : it will manage (open/close) the window for you ; using the **chrome app mode** ; if the websocket brokes : window & server will shutdown (close the window ; the server will shutdown, close the server ; the window will shutdown). You (or your clients) will not see a difference with classical GUI !

In **server mode** : it will act as a classical web server ; and you can use as many clients/browsers as you want, from localhost or from anywhere else. Closing a socket ; just close the socket ;-). It can be hosted on the web, as long as the provider service use python3.

[More on window/server differences](https://github.com/manatlan/wuy/blob/master/wiki/diff.md)

In all cases : it will be pretty easy to produce/freeze an executable (windows/linux/apple) (one file with all html/js embedded), using [pyinstaller/windows](https://github.com/manatlan/wuy/blob/master/BUILD.bat). And share your _super python tool_ to the world (wuy comes with [its own freezer (a wuy app !)](https://github.com/manatlan/wuy/tree/master/examples/wuy_freezer))! And in the future : android !

Since 0.6; **wuy** provide a js method to [proxify http requests](https://github.com/manatlan/wuy/blob/master/wiki/proxify.md), to avoid CORS troubles.

Since 0.9; **wuy** provide get/set methods on client/server side to store/retrieve key/value pairs in a json file on serverside.

It's, a little bit, the same thing as [python eel](https://github.com/ChrisKnott/Eel).



**TODOs**:
* In the future : [cefpython3](https://github.com/cztomczak/cefpython) will be the platform of choice for running **wuy** apps on **android**/iphone (when [cef](https://bitbucket.org/chromiumembedded/cef/issues/1991/add-android-support) and [cefpython3](https://github.com/kivy-garden/garden.cefpython/issues/8) will be ready). Currently **wuy** works with cefpython3, on linux/windows/apple. It's working in the [unittests suite](https://github.com/manatlan/wuy/blob/master/wuy_tests.py) (you can already try ; modify _wuy.py_, replace ChromeApp() by ChromeAppCef())
* Write docs & examples

## The simplest example
This is _the hello world_ of **wuy**

```python
import wuy

class helloWorld(wuy.Window):
    """ <button onclick="wuy.beep()">BEEP</button> """
    size=(100,100)

    def beep(self):
        print("\a BEEP !!!")

helloWorld()
```

## To Test/Run

Download [the zip from here](https://github.com/manatlan/wuy/archive/master.zip)

    $ pip3 install aiohttp winreg
    $ python3 -u an_app.py

(_winreg_ is needed for windows only)

It's an app !

For a regular server example (many clients from anywhere on web)

    $ python3 -u a_server.py

## To Use

Install the lib :

    $ pip3 install wuy winreg

(_winreg_ is needed for windows only)

**And follow the [official tuto](https://github.com/manatlan/wuy/blob/master/wiki/README.md)**.


## See Examples

* [an_app](https://github.com/manatlan/wuy/blob/master/an_app.py) : all kind of tricks
* [an_app2](https://github.com/manatlan/wuy/blob/master/an_app2.py) : an input box
* [an_app3](https://github.com/manatlan/wuy/blob/master/an_app3.py) : an alert box with autodeclared js vars (chain'able windows)!
* [an_app4](https://github.com/manatlan/wuy/blob/master/an_app4.py) : using sync & async rpc calls !
* [an_app5](https://github.com/manatlan/wuy/blob/master/an_app5.py) : using async aiohttp.get (request content from web) & FULLSCREEN mode
* [an_app6](https://github.com/manatlan/wuy/blob/master/an_app6.py) : progress bars (async)
* [an_app7](https://github.com/manatlan/wuy/blob/master/an_app7.py) : html is inside the docstring ! (**SIMPLEST**)
* [an_app8](https://github.com/manatlan/wuy/blob/master/an_app8.py) : just to show how to organize your code when app's expand.
* [an_appVuejs](https://github.com/manatlan/wuy/blob/master/an_appVuejs.py) : using a vuejs app (and  a continuous push from server to client)
* [a_server](https://github.com/manatlan/wuy/blob/master/a_server.py) : a server, tchat service for multiple clients
* [an_upload](https://github.com/manatlan/wuy/blob/master/an_upload.py) : a drag'n'drop upload to server example.

Don't forget to have a look at [real examples](https://github.com/manatlan/wuy/tree/master/examples) too (real apps for real life)

A big real life app : See [jBrout3](https://github.com/manatlan/jbrout3) (in development), it's the rebirth of the good old py2/gtk app : [jBrout2](https://jbrout.manatlan.com). It use vuejs/vuex for front, and the jbrout's lib to manage photos.


## Doc

Use [wuy.Window](https://github.com/manatlan/wuy/blob/master/wiki/api_py_window.md) for app (which open/manage a chrome window app)

Use [wuy.Server](https://github.com/manatlan/wuy/blob/master/wiki/api_py_server.md) for classic http/ws servers

And client side (javascript), in all cases : use [wuy.js](https://github.com/manatlan/wuy/blob/master/wiki/api_js.md)

[Learn More on differences between this two class](https://github.com/manatlan/wuy/blob/master/wiki/diff.md)


changelog
1.4.4.0 (02/07/2019)
 - FIX: replacing doc was bugged in some complex cases
1.4.3.0 (01/07/2019)
 - better release (pytest zzz_OldTest are ok now)
1.4.2.0 (01/07/2019)
 - doc is replacable (can embed vars)
1.4.1.0 (01/07/2019)
 - FIX: "server is down" & CTRL-C exception handling in async (pytest zzz_OldTest are flaky)
1.4.0.0 (10/06/2019)
 - use httpcore 0.4.0
1.3.2.3 BETA: (4/06/2019)
 - require httpcore==0.3.0
1.3.2.2 BETA: (28/05/2019)
 - better cookiestore
1.3.2.1 BETA: (28/05/2019)
 - FIX: parallelizm with begin/end
1.3.1.1 BETA: (27/05/2019)
 - console output: better display in parallelism mode
1.3.1.0 BETA: (27/05/2019)
 - cookies are stored in env, so parallelism works as excpected (each file has his cookiejar)
1.3.0.2 BETA: (27/05/2019)
 - fix: env are cloned when parallelizm
 - fix: req.info status_code is now a real int
1.3.0.1 BETA: (27/05/2019)
 - fix: request was broken for some type of body
1.3.0.0 BETA: (26/05/2019) "httpcore version 0.3.0"
 - use Tom Christie's httpcore 0.3.0 (for sync/async features)
 - option --p for parallelism files
 - reqman as a module : now a COROUTINE (reqman.testContent( yml: str, env: {}) -> MainReponse)
 - FIX: header's values are stringified now
1.2.0.0: (22/05/2019)
 - reqman as a module : now a simple reqman.testContent( yml: str, env: {}) -> MainReponse
1.1.8.0: (18/05/2019) 
 - html escape test name
 - display doc line under the test for the reqman.html
1.1.7.0: (18/05/2019) 
 - module.main now returns something that is usefull when used as a lib
1.1.6.0: (17/05/2019) 
 - better handle NonPlayable test (coz unresolved params) .. should be ok now !
1.1.5.0: (16/05/2019) 
 - FIX: reqman run as console script again ;-)
1.1.4.0: (16/05/2019) 
 - reqman is now a module and a console script (reqman)
1.1.3.0: (16/05/2019) 
 - ability to send bytes in the body (using a method with return bytes, ex: "body: <<|giveMeBytes>>")
 - handle None value as "null" ("" before)
1.1.2.6: (16/05/2019) 
 - just a correct README.md
1.1.2.5: (16/05/2019) 
 - type hints (pyright under vscode)
 - fix: yaml warning without loader
 - fix: another unresolved params won't break the process
1.1.2.4 (27/03/2019)
 - FIX: xml output in html wasn't escaped since 1.1.2.1
1.1.2.3 (25/03/2019)
 - FIX: yaml encoding detection (utf8|cp1252)
1.1.2.2 (22/03/2019)
 - FIX: pytest html output status with "MOCK"
1.1.2.1 (22/03/2019)
 - FIX: html output non-encoded non-ascii chars
1.1.2.0 (21/03/2019)
 - new "break" command, to forget following commands in a file
1.1.1.0 (18/03/2019)
 - FIX: save and path break (unresolved param)
 - FIX: resolve param when in a param resolved in body
 - better handle NonPlyable test (coz unresolved params)
 - pytests 98%cov
1.1.0.1 (18/03/2019)
 - FIX: test that foreach section is a list
1.1.0.0 (15/03/2019)
 - better (py)tested (97%cov)
 - unresolved param non break the tests now (html output is created)
 - py38 deprecated warning removed
1.0.1.2 (10/03/2019)
 - expose internal things for the new (py)tests purpose only
1.0.1.1 (08/03/2019)
 - add optionnal doc'string per request, which is displayed in html output
1.0.1.0 (19/02/2019)
 - it's possible to test json str size now !
1.0.0.0 (19/07/2018)
 - first release ;-) no change at all
0.9.9.20 (16/07/2018)
 - fix: when None is present in <,<=,>,>= -> False
0.9.9.19 (16/07/2018)
 - fix: no html response output when http ko
0.9.9.18 (16/07/2018)
 - fix: response timeout None -> response timeout
0.9.9.17 (16/07/2018)
 - fix: response timeout None -> ""
0.9.9.16 (16/07/2018)
  - win patterns
  - save inherits
  - better html display when not callable
  - tests++
  - ★★
0.9.9.15 (15/07/2018)
  - avoid proc to be bad named - "allez la france!!"
0.9.9.14 (15/07/2018)
  - return of dynamic call++
0.9.9.13 (14/07/2018)
  - new yaml feed'er, and foreach in foreach works
  - control keywords (more robust)
0.9.9.12 (14/07/2018)
  - dynamic foreach & more robust (but not foreach in foreach)
0.9.9.11 (14/07/2018)
  - remove params as list -> replaced by foreach
0.9.9.10 (13/07/2018)
  - assert params is a list or a dict : not str !
0.9.9.9 (13/07/2018)
  - params as list -> foreach params : it will request for each param in the list
  - regression: no more methods/param in call line (not really useful, and not used (by me or us))
0.9.9.8 (13/07/2018)
  - f-string removal to have compatility with py3
0.9.9.72 (12/07/2018)
  - just for create a new pip package
0.9.9.71 (12/07/2018)
  - minor changes (output troubles in console) (and WHL/wheel release)
0.9.9.7 (11/07/2018)
  - ctrl-c render the html output (again ;-)
  - test system has been completly rewritten (comparators now works for all tests, and inside list too)
  - use f-string now
0.9.9.6 (08/07/2018)
  - ctrl-c break the html output (for now)
  - FIX: a Test() trouble (exception ignored)
0.9.9.52 (05/07/2018)
  - Ctrl-C render html anyway
0.9.9.5 (03/07/2018)
  - better comparators, try/catch better, html outpu better
0.9.9.4 (02/07/2018)
  - compare full json restult with content
0.9.9.3 (02/07/2018)
  - comparators (for json.* & status) are in the box !!
0.9.9.2 (01/07/2018)
  - HTML RENDERING : None as ""
0.9.9.1 (29/06/2018)
  - first release with all features in (next will be 1.0.0.0)


wuy_ut.py
# -*- coding: utf-8 -*-
import wuy,sys,asyncio,os
import unittest

# Officiel unit tests (more than coverage 80% of code (cef works too), windows-mode only)
#TODO: may block under windows (but each test works independantly) ... should find a solution ! It works as expected unde linux only (mac ?)

ONLYs=[]
def only(f): # decorator to place on tests, to limit usage to only theses ones
    ONLYs.append(f.__name__)
    return f

JSTEST=25 # there are 'JSTEST' js tests in class vv, to test with app-mode and cefpython3 !
class UnitTests(wuy.Window):
    """
    <meta charset="utf-8" />

    I am 
    <script>
    document.write(wuy.iam);

    document.addEventListener("init", async function() {

        await wuy.report("declared var", wuy.val=="mémé" ,"")

        await testRequestPath("/unknown", r=>r.status==404 )
        await testRequestPath("/1", r=>r.status==200 && r.headers["content-type"].includes("text/html") )
        await testRequestPath("/2", r=>r.status==200 && r.headers["content-type"].includes("application/octet-stream") )
        await testRequestPath("/11", r=>r.status==201 && r.headers["content-type"].includes("text/html") )
        await testRequestPath("/12", r=>r.status==201 && r.headers["content-type"].includes("application/octet-stream") )
        await testRequestPath("/20", r=>r.status==202 && r.headers["content-type"].includes("text/plain") )
        await testRequestPath("/30", r=>r.status==203 && r.headers["content-type"].includes("text/plain") )

        await testBadCall("Test unknown proc",()=>wuy.unknown() );
        await testBadCall("Test call proc with bad number of params",()=>wuy.call(1) );
        await testBadCall("Test async call proc with bad number of params",()=>wuy.acall(1) );

        await wuy.report("Test good sync call", "abc" == await wuy.call('a','b','c') ,"")
        await wuy.report("Test good async call", "abc" == await wuy.acall('a','b','c') ,"")
        await wuy.report("Test good sync call", "abc" == await wuy.call('a','b') ,"")
        await wuy.report("Test good async call", "abc" == await wuy.acall('a','b') ,"")

        //---------------------------------------------------------
        // test a method which send a event

        r=await wuy.testEmit()
        wuy.on("pyEvent",async function(a,b,c) {
            await wuy.report("Test py/emit event recepted",a=="a" && b=="b" && c=="c","")
        })
        await wuy.report("Test py/emit return",r=="ok","")

        //---------------------------------------------------------
        // test wuy.emit() (not usefull in windows mode !)

        wuy.on("jsEvent",async function(a,b,c) { // not called in windows mode
            await wuy.report("jsEvent not received (normal)",False,"")  // it's bad !
        })
        var p=await wuy.emit("jsEvent",1,2,3)
        await wuy.report("Test js/emit return good things", p.length==3 && p[0]==1 && p[1]==2 && p[2]==3 ,"")

        //---------------------------------------------------------

        var content=await wuy.testWuyRequest()
        await wuy.report("Test wuy.request GET", content.includes("wuy.js") ,"")

        var content=await wuy.testWuyRequestPost()
        await wuy.report("Test wuy.request POST", content.includes("'wuy'") ,"")

        await testBadCall("Test wuy.request Error",()=>wuy.testWuyRequestError() )

        var content=await fetch("UnitTests.html").then( x=>x.text())
        await wuy.report("Test classic fetch", content.includes("wuy.js") ,"")

        //TODO: the givent port is arbitary ... change to self._port !
        var content=await wuy.fetch("http://localhost:8080/UnitTests.html").then( x=>x.text())  // need to have FULL URL !
        await wuy.report("Test wuy.fetch", content.includes("wuy.js") ,"")

        await testDivError("Test error",()=>wuy.testError() );
        await testDivError("Test async error",()=>wuy.atestError() );

        window.close()
    });

    async function testRequestPath(p,cbt) {
        var r=await fetch(p)
        var h={};
        for(var i of r.headers) h[ i[0] ] = i[1]
        var rep={status:r.status, headers:h, content: await r.text()}

        wuy.report(`Test path "${p}"`, cbt(rep), "Tests:"+cbt+" with:"+JSON.stringify(rep))
    }

    async function testBadCall(p,cbt) {
        try {
            await cbt()                 // raise an exception !
            await wuy.report(p,false,"")
        }
        catch(e) {wuy.report(p,true,e)}
    }

    async function testDivError(p,cbt) {
        try {
            await cbt()
            await wuy.report(p, False ,"")
        }
        catch(e) {
            await wuy.report(p, (""+e).includes("division by zero") ,"")
        }
    }


    </script>

    """
    size=(100,100)

    def init(self):
        self.tests=[]

    def report(self,libl,val,info):
        self.tests.append( (val,libl,info) )

    #===========================================================================
    def request(self,req):    # async or not
        if req.path=="/1":
            return "text body"   # 200, text/html
        elif req.path=="/2":
            return b"text body"  # 200, application/octect-stream
        elif req.path=="/11":
            return wuy.Response(201,"text body")    # text/html
        elif req.path=="/12":
            return wuy.Response(201,b"text body")   # application/octect-stream
        elif req.path=="/20":
            return wuy.Response(202,"text body","text/plain")    # text/plain
        elif req.path=="/30":
            return wuy.Response(203,"text body",{"content-type":"text/plain"})
        if req.path=="/post" and req.method=="POST":
            return "say hello to '%s'"%req.body   # 200, text/html

    def call(self,a,b,c="c"):
        return "%s%s%s" % (a,b,c)

    async def acall(self,a,b,c="c"):
        return "%s%s%s" % (a,b,c)

    def testEmit(self):
        self.emit("pyEvent","a","b","c")
        return "ok"

    async def atestError(self):
        a=12/0
        return "ok" 

    def testError(self):
        a=12/0
        return "ok" 

    async def testWuyRequest(self):
        return (await wuy.request("http://localhost:%s/UnitTests.html" % self._port)).content

    async def testWuyRequestPost(self):
        return (await wuy.request("http://localhost:%s/post" % self._port,"wuy")).content

    async def testWuyRequestError(self):
        return await wuy.request("hvfgdfdsgfdsgdf")

def reportJS(ll,txt):
    print("#"*79,txt)
    for test,libl,info in ll:
        print( test and "ok:" or "KO:",libl, "" if test is True else info)
    print("#"*79)

class TestWuy(unittest.TestCase):

    def test_isFreePort(self):
        ll=[wuy.isFree("localhost",p) for p in [22,23,445]]
        self.assertTrue( False in ll)

    def test_path(self):
        self.assertFalse(hasattr(sys,"_MEIPASS"))
        self.assertEqual(wuy.path("jo"),"jo")

    def test_pathFrozen(self):
        sys._MEIPASS="kiki"
        self.assertEqual(wuy.path("jo").replace("\\","/"),"kiki/jo")
        delattr(sys,"_MEIPASS")
        self.assertFalse(hasattr(sys,"_MEIPASS"))

    def test_getname(self):
        self.assertEqual( wuy.getname("jo"),"jo" )
        self.assertEqual( wuy.getname("jo.html"),"jo" )
        self.assertEqual( wuy.getname("jim/jo"),"jim.jo" )
        self.assertEqual( wuy.getname("jim/jo.html"),"jim.jo" )

    def test_a_window_render_html(self):
        class aeff(wuy.Window):
            size=(100,100)
            def init(self):
                asyncio.get_event_loop().call_later(2, self.exit)

        if os.path.isfile("web/aeff.html"): os.unlink("web/aeff.html")
        aeff()
        self.assertTrue( os.path.isfile("web/aeff.html") )
        os.unlink("web/aeff.html")

    def test_a_double_window(self):
        class aeff(wuy.Window):
            "test double open"
            size=(100,100)
            def init(self):
                asyncio.get_event_loop().call_later(2, self.exit)

        aeff()
        aeff()

    def test_a_server(self):
        class saeff(wuy.Server):
            "I'm a server"
            def init(self):
                asyncio.get_event_loop().call_later(2, self.exit)
        saeff()
        self.assertEqual( len(wuy.currents),1)  # there was one instance

    def test_2_server(self):
        class saeff1(wuy.Server):
            "I'm a server, and the quitter"
            def init(self):
                asyncio.get_event_loop().call_later(2, self.exit)
        class saeff2(wuy.Server):
            "I'm a server, and I will killed by saeff1"
            pass
        wuy.Server.run()
        self.assertEqual( len(wuy.currents),2) # there were 2 instances

    # @only
    def test_a_window(self):                # <--- main tests here
        ut=UnitTests(log=True,val="mémé",iam="local chrome")

        reportJS(ut.tests,"main tests on local chrome app-mode")

        self.assertEqual( len([ok for ok,*_ in ut.tests if ok]),JSTEST)        # 23 tests ok

    def test_cef_if_present(self):
        import pkgutil 
        if pkgutil.find_loader("cefpython3"):
            try:
                old=wuy.ChromeApp
                wuy.ChromeApp=wuy.ChromeAppCef

                ut=UnitTests(log=True,val="mémé",iam="cef")
                reportJS(ut.tests,"main tests with cefpython3")
                self.assertEqual( len([ok for ok,*_ in ut.tests if ok]),JSTEST)        # 23 tests ok

            finally:
                wuy.ChromeApp=old
        else:
            print("***WARNING*** : cefpython3 not present, no tests !")

    # def test_a_windows(self):
    #     s = 'hello world'
    #     self.assertEqual(s.split(), ['hello', 'world'])
    #     # check that s.split fails when the separator is not a string
    #     with self.assertRaises(TypeError):
    #         s.split(2)


if __name__=="__main__":
    if ONLYs:
        print("*** WARNING *** skip some tests !")
        def load_tests(loader, tests, pattern):
            suite = unittest.TestSuite()
            for c in tests._tests:
                suite.addTests( [t for t in c._tests if t._testMethodName in ONLYs] )
            return suite

    unittest.main( )
